### PR TITLE
Initial patch for FlowContainer.

### PR DESCRIPTION
**Pull request by djaiswal83**
Contains plumbing for in-memory KeyStore.

### CODE REVIEW

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+
+
+  private FlowContainer(final Path projectDirPath, final Props props)
+          throws IOException {
+    this.projectDir = projectDirPath.toFile();
+
+    // Create Azkaban Props Map
+    this.azKabanProps = props;
+    // Setup global props if applicable
+    final String globalPropsPath = this.azKabanProps.getString("executor.global.properties", null);
```
Comment (aditya1105) [inline azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:115]: Can be converted to constant

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+    this.jobLogNumFiles = this.azKabanProps.getInt("job.log.backup.index", 4);
+    this.validateProxyUser = this.azKabanProps.getBoolean("proxy.user.lock.down", false);
+    this.jobTypeManager =
+        new JobTypeManager(
+            this.azKabanProps.getString(AzkabanExecutorServer.JOBTYPE_PLUGIN_DIR,
+                PluginManager.JOBTYPE_DEFAULTDIR),
+            this.globalProps, getClass().getClassLoader());
+    // Fetch keyStore props and use it get the KeyStore, put it in JobTypeManager
+    Props keyStoreLoadProps = this.jobTypeManager.getKeyStoreLoadProps();
+    if (keyStoreLoadProps != null) {
```
Comment (aditya1105) [inline azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:145]: If keyStoreLoadProps is null, do we want to throw an exception or consume it? If we intend to throw an exception probaly early check like below will be helpful
if (null == keyStoreLoadProps) {
 // throw exception
}

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+      }
+      jobTypeManager.setKeyStore(keyStore);
+    }
+  }
+
+  public static void main(final String[] args) throws IOException, ExecutorManagerException {
+    // Redirect all std out and err messages into log4j
+    StdOutErrRedirect.redirectOutAndErrToLog();
+    // Get all the arguments
+    final String execIdStr = args[0];
```
Comment (aditya1105) [inline azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:160]: Argument length check is essential before calling indexes 0 and 1

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+    int execId = 0;
+    try {
+      execId = Integer.parseInt(execIdStr);
+    } catch (NumberFormatException ne) {
+      System.out.printf("Execution ID %s is invalid %n", args[0]);
+    }
+    System.out.printf("Execution ID = %d%n", execId);
+
+    // Setup work directories
+    final Path currentWorkingDir = Paths.get("").toAbsolutePath();
```
Comment (aditya1105) [inline azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:172]: These and below two paths should be config based

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+    System.out.println("moving projectDir:" + projectZipName + ": to " +
+        projectZipPath);
+    Files.move(Paths.get(projectZipName), projectZipPath);
+    // TODO : revisit this logic,
+    // Unzip the project zip
+    FlowContainer.unzipFile(projectZipPath.toString(), projectDirPath);
+
+    // TODO : Deepak - throw away after directory setup is final.
+    Path pwd = Paths.get("").toAbsolutePath();
+    System.out.println("Deepak : Dumping dir structure for = " + pwd.toString());
```
Comment (aditya1105) [inline azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:191]: Is this a temporary code?

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        new FlowContainer(projectDirPath, props);
+
+    // Use some execId to execute the flow
+    flowContainer.submitFlow(execId);
+  }
+
+  // Set Azkaban Props
+  private static Props setAzkabanProps(final Path jobtypePluginPath) {
+    final Map<String, String> propsMap = new HashMap<>();
+    propsMap.put(AzkabanExecutorServer.JOBTYPE_PLUGIN_DIR,
```
Comment (aditya1105) [inline azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:215]: Again this should be be config based

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    StdOutErrRedirect.redirectOutAndErrToLog();
+    // Get all the arguments
+    final String execIdStr = args[0];
+    final String projectZipName = args[1];
+    // Process Execution ID.
+    int execId = 0;
+    try {
+      execId = Integer.parseInt(execIdStr);
+    } catch (NumberFormatException ne) {
+      System.out.printf("Execution ID %s is invalid %n", args[0]);
```
Comment (aditya1105) [inline azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:167]: Do we intend to continue or throw exception

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+    try {
+      this.flowFuture.get();
+    } catch (final InterruptedException ie) {
+      ie.printStackTrace();
+    } catch (final ExecutionException ee) {
+      ee.printStackTrace();
+    }
+  }
+
+  private static void unzipFile(final String zipPath,
```
Comment (aditya1105) [inline azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:273]: Is this a temporary code?

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+    final ExecutionOptions options = flow.getExecutionOptions();
+    if (options.getPipelineExecutionId() != null) {
+      final int pipelinedExecId = options.getPipelineExecutionId();
+      watcher = new RemoteFlowWatcher(pipelinedExecId, this.executorLoader);
+    }
+
+    // TODO : figure out the metrics
+    final MetricsManager metricsManager = new MetricsManager(new MetricRegistry());
+    final CommonMetrics commonMetrics = new CommonMetrics(metricsManager);
+    final ExecMetrics execMetrics = new ExecMetrics(metricsManager);
```
Comment (aditya1105) [inline azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:248]: ExecMetrics have @inject, can guice be utilized here?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunner.java [hunk 10]
```diff
@@ -71,10 +71,7 @@
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import java.io.File;
 import java.io.IOException;
-import java.security.AccessControlContext;
-import java.security.AccessController;
-import java.security.PrivilegedExceptionAction;
-import java.security.ProtectionDomain;
+import java.security.*;
```
Comment (aditya1105) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunner.java:74]: It is better to resolve * with actual imports

File: azkaban-exec-server/src/main/java/azkaban/execapp/JobRunner.java [hunk 11]
```diff
@@ -742,6 +749,19 @@ private Status prepareJob() throws RuntimeException {
 
       try {
         this.job = this.jobtypeManager.buildJobExecutor(this.jobId, this.props, this.logger);
+        if (keyStore != null &&
```
Comment (aditya1105) [inline azkaban-exec-server/src/main/java/azkaban/execapp/JobRunner.java:752]: If keystore==null, there is no exception thrown

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.zip.ZipFile;
+import javax.sql.DataSource;
+import org.apache.commons.dbutils.QueryRunner;
+import org.apache.log4j.Logger;
```
Comment (arvindpruthi) [inline azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:74]: Should you use log4j v2 in the new file?

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java (hunk 1 above)
Comment (arvindpruthi) [reply azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:115]: +1

File: azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+    final CommonMetrics commonMetrics = new CommonMetrics(metricsManager);
+    final ExecMetrics execMetrics = new ExecMetrics(metricsManager);
+    final FlowRunner flowRunner = new FlowRunner(flow, this.executorLoader,
+        this.projectLoader, this.jobTypeManager, this.azKabanProps, null,
+            null, commonMetrics, execMetrics);
+    flowRunner.setKeyStore(this.keyStore);
+    flowRunner.setFlowWatcher(watcher)
+        .setJobLogSettings(this.jobLogChunkSize, this.jobLogNumFiles)
+        .setValidateProxyUser(this.validateProxyUser)
+        .setNumJobThreads(20);
```
Comment (arvindpruthi) [inline azkaban-exec-server/src/main/java/azkaban/container/FlowContainer.java:256]: Should this come from a config parameter?

**Review by arvindpruthi — COMMENTED** [review]
Overall documentation needs to improve a lot. At the very least every class and every function should have some basic documentation to describe its purpose. I understand this is a WIP at the moment. But we should take this opportunity to also document, otherwise poorly documented code that we are touching.

**Pull request closed without merging**